### PR TITLE
Dart: Add UInt8List support, allowing for type:string format:binary definitions

### PR DESCRIPTION
**Pull request by 0xNF**
## Summary
tl;dr `UInt8List` support, which Dart-Dio has, has been added to the regular Dart2 generator.

This PR fixes issue #12161.

### Changes to the Generators:

To support UInt8List in the generator, Typed Data imports had to be added. Additionally, some OpenAPI data types have been switched to Uint8List.

* [AbstractDartCodeGen.java]:  added `Uint8List/dart:typed_data` to the imports.
* [AbstractDartCodeGen.java]: Changed `typeMapping` of {file, binary} to `Uint8List` away from `MultiPartFile`
* [AbstractDartCodeGen.java]:  Copied Multipart/Uint8list differentiation in `postProcessOperationsWithModels` from DartDioClientCodegen.java
*  [dart2/apilib.mustache]: Added `import 'dart:typed_data';`

### Changes to the API classes

To support Uint8List in the generated API output, parts of the *api.dart files had to be refactored. Presently Dart2 relies on passing utf8 decoded response byte strings through its API calls, but this leads to errors because not all response objects are json strings. To handle this, I refactored many of the calls to take Uint8List's straight from `response.bodyBytes`. Decoding into a `String` object doesn't happen until it needs to, which is right before passing the object down into the Json Decoding section.

* [api_client.dart]: deserialize and deserializeAsync now take Uint8List as their default input
* [api_helper.dart]: removed _decodeBodyBytes in favor of just using `response.bodyBytes`
* [api_helper.dart]: added `decodeError` method to deal with the byte response, which decodes the resposne into a utf-8 String.

To confirm this change, use the following specfile:

<details>
  <summary>specfile</summary>

```yaml
openapi: 3.0.3
info:
  version: "1.1"
  title: Dart Uint8list Demo
servers:
  - url: "localhost"
    variables:
      host:
        default: localhost
paths:
  /getData:
    get:
      operationId: GetData
      description: "Should return a Uint8List of bytes"
      responses:
        "200":
          description: Binary data
          content:
            application/octet-stream:
              schema:
                type: string
                format: binary
    put:
      operationId: PutData
      description: "Should push a Multipart File"
      requestBody:
        content:
          multipart/form-data:
            schema:
              type: object
              properties:
                fileContent:
                  type: string
                  format: binary
      responses:
        "200":
          description: OK
          content:
            application/json:
              schema:
                type: string
  /item:
    get:
      operationId: GetItem
      description: "Should return an Item"
      responses:
        "200":
          description: items
          content:
            application/json:
              schema:
                $ref: "#components/schemas/item"
components:
  schemas:
    Item:
      type: object
      description: "Some json thing"
      required:
        - intField
        - doubleField
        - stringField
        - dateField
        - timeField
        - dateTimeField
        - boolField
        - b64Field
        - intArrayField
      properties:
        intField:
          type: integer
          format: int32
        doubleField:
          type: number
          format: float
        stringField:
          type: string
        dateField:
          type: string
          format: date
        timeField:
          type: string
          format: time
        dateTimeField:
          type: string
          format: date-time
        boolField:
          type: boolean
        b64Field:
          type: string
          format: byte
        intArrayField:
          type: array
          items:
            type: integer
            format: int32
```
</details>

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.1.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@jaumard  @josh-burton @amondnet @sbu-WBT @kuhnroyal @agilob @ahmednfwela

### CODE REVIEW

File: samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/api/pet_api.dart [hunk 1]
```diff
@@ -10,23 +10,24 @@
...
-
 class PetApi {
   PetApi([ApiClient? apiClient]) : apiClient = apiClient ?? defaultApiClient;
 
   final ApiClient apiClient;
 
   /// Add a new pet to the store
   ///
-  /// 
+  ///
```
Comment (ahmednfwela) [inline samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/api/pet_api.dart:20]: please avoid whitespace changes since they pollute the PR

File: samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/api/pet_api.dart [hunk 2]
```diff
@@ -10,23 +10,24 @@
...
   ///
   /// Note: This method returns the HTTP [Response].
   ///
   /// Parameters:
   ///
   /// * [Pet] pet (required):
   ///   Pet object that needs to be added to the store
-  Future<Response> addPetWithHttpInfo(Pet pet,) async {
+  Future<Response> addPetWithHttpInfo(
+    Pet pet,
```
Comment (ahmednfwela) [inline samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/api/pet_api.dart:29]: whitespace changes here too

**Review by ahmednfwela — CHANGES_REQUESTED** [review]
there are alot of whitespace changes here that are polluting the PR, it would be great if the PR had the bare minimum changes

**Comment by 0xNF** [comment]
When I get home I'll try to fix it, but this I didn't touch any of those files by hand. I only ran the generate samples script. I didn't even open these in my editor.

File: samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/api/pet_api.dart (hunk 2 above)
Comment (0xNF) [reply samples/openapi3/client/petstore/dart2/petstore_client_lib/lib/api/pet_api.dart]: The newest version of this branch should have corrected the extraneous whitespace.

File: modules/openapi-generator/src/main/resources/dart2/api_client.mustache [hunk 3]
```diff
@@ -134,19 +134,25 @@ class ApiClient {
...
 
+    // if the expected target type is a byte-array, then just return the response bytes as-is
+    if (targetType == 'Uint8List') {
+      return responseBytes;
+    }
+
+    final decodedString = const Utf8Decoder().convert(responseBytes);
     // If the expected target type is String, nothing to do...
     return targetType == 'String'
       ? json
```
Comment (ahmednfwela) [inline modules/openapi-generator/src/main/resources/dart2/api_client.mustache:154]: shouldn't this return `decodedString`?

**Review by ahmednfwela — CHANGES_REQUESTED** [review]

File: modules/openapi-generator/src/main/resources/dart2/api_client.mustache (hunk 3 above)
Comment (0xNF) [reply modules/openapi-generator/src/main/resources/dart2/api_client.mustache]: Eagle-eye'd catch there. `json` isn't even a variable in the new method anymore, my mistake New push of the branch has it returning `decodedString` now.

**Comment by ahmednfwela** [comment]
@kuhnroyal @wing328 Can you check why the tests were passing here when there was a compile error for the generated code?

**Comment by ahmednfwela** [comment]
@0xNF did you make any changes in the java files? I can't see them in the PR

**Comment by 0xNF** [comment]
The Java changes seem to have gotten lost. Sorry this PR is kinda messy. I'm flipping between too many branches rn.

**Comment by wing328** [comment]
> @kuhnroyal @wing328 Can you check why the tests were passing here when there was a compile error for the generated code?

The test did run and passed. Does it has anything to do with Dart version?

```
      - uses: dart-lang/setup-dart@v1
        with:
          sdk: 2.14.0
```

That's what we use in the CI.

**Comment by ahmednfwela** [comment]
dart 2 uses 2.12 in the generated code, which 2.14 should be able to run

**Comment by wing328** [comment]
Can you please paste the compilation error here as well? I'll try to repeat that locally.

**Comment by 0xNF** [comment]
It's probably because I force pushed the update to my own branch, and the tests didn't re-run here.

**Comment by wing328** [comment]
> It's probably because I force pushed the update to my own branch, and the tests didn't re-run here.

To trigger the run, please update the samples and the Github action (dart job) will run again.

**Comment by 0xNF** [comment]
Should I be regenerating all samples? Up til now I've just been regenerating the relevant samples for my changes (Dart). But then the CI complains.

**Comment by zmoshansky** [comment]
@0xNF @wing328  Is there any movement on this? It looks like this PR is functionally ready to land. Just needing some clean-up.

Currently, I'm facing an issue around this as passing a `UInt8List`  to a generated API endpoint causes the following error:
```
DioError [DioErrorType.other]: Bad state: No serializer for 'Uint8List'.
```

Relevant Code (Generated API's model.dart):
```
    yield serializers.serialize(
      object.appBinaryData,
      specifiedType: const FullType(Uint8List),
    );
```

**Comment by ahmednfwela** [comment]
@zmoshansky this PR is for `dart` generator, and it seems like you are using the `dart-dio` generator, please create a new Issue with reproducible sample
